### PR TITLE
test: UT and annotation for Bspline class

### LINKED ISSUE #748: We lack a unit test for class Bspline in Module_base
## Describe Current Status and Possible Solution

We lack a unit test for class Bspline in Module_base.

## Additional Context

Currently, this class is the application of methods in "J. Chem. Phys. 103, 8577 (1955)" to calculate structure factor. It is not used for B-spline interpolation.

### PR DESCRIPTION
**Pull request by hongriTianqi**

### CODE REVIEW

**Review by Qianruipku — APPROVED** [review]

File: source/module_base/math_bspline.cpp [hunk 1]
```diff
@@ -1,64 +1,65 @@
...
+    Dx = 1.0;
+}
+Bspline::~Bspline()
+{
+    if (bezier != NULL)
+        delete[] bezier;
+}
 
-    void Bspline::init(int norderin, double Dxin, double xiin)
+void Bspline::init(int norderin, double Dxin, double xiin)
```
Comment (caic99) [inline source/module_base/math_bspline.cpp:21]: Is it better to move contents of `init` into its constructor? Same for `cleanup` to destructor.

File: source/module_base/math_bspline.cpp [hunk 2]
```diff
@@ -1,64 +1,65 @@
...
+        bezier[i] = 0;
     }
+}
 
-    void Bspline::cleanp()
-    {
-        delete[] bezier;
-        bezier = NULL;
-    }
+void Bspline::cleanp()
```
Comment (caic99) [inline source/module_base/math_bspline.cpp:37]: `clean_up` typo?

File: source/module_base/math_bspline.h [hunk 3]
```diff
@@ -4,57 +4,58 @@
...
-//   bp.init(10,0.7,2); //Dx = 0.7, xi = 2
-//   bp.getbslpine(0.5); //x = 0.5
-//   cout<<bp.bspline_ele(3)<<endl; //print M_ik[xi+3*Dx+x]: M_i[10](4.6) 
-//
+/**
+ * @brief A class to treat Cardinal B-spline interpolation.
+ *
+ * @author qianrui created 2021-09-14
+ * @details
+ * Math:
```
Comment (caic99) [inline source/module_base/math_bspline.h:12]: Maybe adding a reference for the source of the formulas?

File: source/module_base/math_bspline.h [hunk 4]
```diff
@@ -4,57 +4,58 @@
...
 
-    public:
-        Bspline();
-        ~Bspline();
+    // Init norder, Dx, xi
+    void init(int norderin, double Dxin, double xiin);
 
-        //Init norder, Dx, xi
-        void init(int norderin, double Dxin, double xiin);
+    // delete[] bezier
```
Comment (caic99) [inline source/module_base/math_bspline.h:49]: Please remove redundant codes.

**Review by caic99 — CHANGES_REQUESTED** [review]

File: source/module_base/math_bspline.cpp (hunk 1 above)
Comment (Qianruipku) [reply source/module_base/math_bspline.cpp:21]: We want to reuse the same object of this class. Then it is necesarry. For examle:
Bspline a;
a.init(10,2,3);
do something
a.cleanup();
a.init(8,1,3);
do something
a.cleanup();

File: source/module_base/math_bspline.cpp (hunk 2 above)
Comment (Qianruipku) [reply source/module_base/math_bspline.cpp:37]: Yes...

File: source/module_base/math_bspline.cpp (hunk 1 above)
Comment (caic99) [reply source/module_base/math_bspline.cpp:21]: > We want to reuse the same object of this class. Then it is necesarry. For examle: Bspline a; a.init(10,2,3); do something a.cleanup(); a.init(8,1,3); do something a.cleanup();

I see. 
However, what is your motivation to reuse the object? Is there any relationship between `a.init(10,2,3);` and `a.init(8,1,3);`, for example the results generated will be used afterwards? Or simply saving space?

If there are no relationship between them(I think it is), I suggest separating them to `Bspline a,b`, for reusing the same object might not be intuitive, and even potentially hazardous.

Further more, the code will be more clear in logic flow while preventing misuse. Afterwards, you only need to do something like `Bspline a(1,2,3); do something;`: no init, no cleanup, no manually allocate, free(!) spaces, all handled with constructor and destructor.

Comment (Qianruipku) [reply source/module_base/math_bspline.cpp:21]: Actually, ABACUS do not need to reuse it. Maybe, you can simplify this class, @hongriTianqi .

Comment (hongriTianqi) [reply source/module_base/math_bspline.cpp:21]: The cleanup function has exactly the same contents as destructor, so I think it is fine to delete it. It is actually not called and used in ABACUS.

Comment (hongriTianqi) [reply source/module_base/math_bspline.cpp:21]: But I think we need to retain the function init. The reason is that here we need to assert that norder is an even and positive integer. I think it is inappropriate to do assertions in a constructor. What's you opinion? @caic99 @Qianruipku

**Pull request closed without merging**